OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs

```diff
@@ -70,7 +70,8 @@ public static bool IsCellBuildable(this World world, CPos cell, ActorInfo ai, Bu
 			}
 
 			// Buildings can never be placed on ramps
-			return world.Map.Ramp[cell] == 0 && bi.TerrainTypes.Contains(world.Map.GetTerrainInfo(cell).Type);
+			return world.Map.Ramp[cell] == 0 && bi.TerrainTypes.Contains(world.Map.GetTerrainInfo(cell).Type) &&
+				!world.ActorMap.IsBuildingBlocked(cell);
```

Comment:
I suggest moving those checks up to a) prevent `IsBuildingBlocked` from having an effect on `AllowInvalidPlacement` and b) save us from searching actors in a cell that is not buildable anyway.
```csharp
			if (!world.Map.Contains(cell))
				return false;

			// Buildings can never be placed on ramps
			if (world.Map.Ramp[cell] != 0 || !bi.TerrainTypes.Contains(world.Map.GetTerrainInfo(cell).Type))
				return false;

			if (bi.AllowInvalidPlacement)
				return true;

			if (world.ActorMap.IsBuildingBlocked(cell))
				return false;

			// Replaceable actors are rare, so avoid initializing state unless we have to
			var checkReplacements = ai != null && ai.HasTraitInfo<ReplacementInfo>();
			HashSet<string> acceptedReplacements = null;

			var foundActors = false;
			foreach (var a in world.ActorMap.GetActorsAt(cell))
			{
				if (a == toIgnore)
					continue;

				// If this is potentially a replacement actor we must check *all* cell occupants
				// before we know the placement is invalid
				// Otherwise, we can bail immediately
				if (!checkReplacements)
					return false;

				foundActors = true;
				foreach (var r in a.TraitsImplementing<Replaceable>())
				{
					if (r.IsTraitDisabled)
						continue;

					if (acceptedReplacements == null)
						acceptedReplacements = new HashSet<string>();

					acceptedReplacements.UnionWith(r.Info.Types);
				}
			}

			// Replacements are enabled and the cell contained at least one (not ignored) actor
			if (foundActors)
			{
				// The cell contains at least one actor, and none were replaceable
				if (acceptedReplacements == null)
					return false;

				// The cell contains at least one replaceable actor, but not of the types we accept
				var foundReplacement = ai.TraitInfos<ReplacementInfo>()
					.Any(r => r.ReplaceableTypes.Overlaps(acceptedReplacements));

				if (!foundReplacement)
					return false;
			}

			return true;
```

Reply:
On a deeper look, I don't think this is correct. `IsBuildingBlocked` is supposed to have an effect on `AllowInvalidPlacement`: the expected behaviour is that those cells should show red on the placement preview, but allow placement anyway (and those cells are simply ignored).

The suggested changes regress the concrete placement behaviour in d2k: hovering the concrete PBOG over a building that that was built on concrete correctly shows red, but hovering the concrete PBOG over a building that was built on rock shows white!

Reply:
In the longer term (i'm not sure its feasible for the upcoming playtest) it would probably be better to use the yellow placement tiles instead of red in this case - red is supposed to mean that the placement is not allowed at all, whereas yellow means it is allowed but those specific cells are going to behave differently.

Reply:
> The suggested changes regress the concrete placement behaviour in d2k: hovering the concrete PBOG over a building that that was built on concrete correctly shows red, but hovering the concrete PBOG over a building that was built on rock shows white!

I actually expected something this, and think it would be better to always show white not red, simply because a full red preview won't block you from placing it down (when you would expect it to reject placement; I think we have an open issue for that). I guess this doesn't matter if we remove `AllowInvalidPlacement` in the long run.

Reply:
My suggestion for a specific `D2kConcrete` could fix that too.

Reply:
Btw, I just remembered what is going on here. Whatever we do in this method is kind of a moot point because of https://github.com/OpenRA/OpenRA/blob/daa8c74c375d88999fa4671cecb45fe4222d2548/OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs#L76-L79
Which means returning false here just results in a red color that is completely ignored.

Reply:
Indeed. That's what I said above 😄